### PR TITLE
[Merged by Bors] - feat: port Data.Set.Prod

### PR DESCRIPTION
**Pull request by kim-em**
fc2ed6f838ce7c9b7c7171e58d78eaf7b438fb0e

- [x] depends on: #1000

### CODE REVIEW

**Comment by semorrison** [comment]
Sorry, I hadn't seen #978. I've merged three proofs from that branch into this.

**Comment by ChrisHughes24** [comment]
bors merge

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#1000~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Set/Prod.lean [hunk 1]
```diff
@@ -0,0 +1,891 @@
...
+
+This file defines the product of sets in `α × β` and in `Π i, α i` along with the diagonal of a
+type.
+
+## Main declarations
+
+* `Set.prod`: Binary product of sets. For `s : Set α`, `t : Set β`, we have
+  `s.prod t : Set (α × β)`.
+* `Set.diagonal`: Diagonal of a type. `Set.diagonal α = {(x, x) | x : α}`.
+* `Set.off_diag`: Off-diagonal. `s ×ˢ s` without the diagonal.
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Prod.lean:20]:
```suggestion
* `Set.offDiag`: Off-diagonal. `s ×ˢ s` without the diagonal.
```

File: Mathlib/Data/Set/Prod.lean [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+#align set.off_diag_nonempty Set.offDiag_nonempty
+
+@[simp]
+theorem offDiag_eq_empty : s.offDiag = ∅ ↔ s.Subsingleton := by
+  rw [← not_nonempty_iff_eq_empty, ← not_nontrivial_iff, offDiag_nonempty.not]
+#align set.off_diag_eq_empty Set.offDiag_eq_empty
+
+alias offDiag_nonempty ↔ _ nontrivial.offDiag_nonempty
+
+alias offDiag_nonempty ↔ _ subsingleton.offDiag_eq_empty
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Prod.lean:567]:
```suggestion
alias offDiag_nonempty ↔ _ Nontrivial.offDiag_nonempty

alias offDiag_nonempty ↔ _ Subsingleton.offDiag_eq_empty
```

File: Mathlib/Data/Set/Prod.lean [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+
+@[simp]
+theorem disjoint_diagonal_off_diag : Disjoint (diagonal α) s.offDiag :=
+  disjoint_left.mpr fun _ hd ho => ho.2.2 hd
+#align set.disjoint_diagonal_off_diag Set.disjoint_diagonal_off_diag
+
+theorem offDiag_inter : (s ∩ t).offDiag = s.offDiag ∩ t.offDiag :=
+  ext fun x => by
+    simp only [mem_offDiag, mem_inter_iff]
+    constructor
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Prod.lean:604]:
```suggestion
    -- Porting note: was `tauto`
    constructor
```

File: Mathlib/Data/Set/Prod.lean [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+#align set.preimage_pi Set.preimage_pi
+
+theorem pi_if {p : ι → Prop} [h : DecidablePred p] (s : Set ι) (t₁ t₂ : ∀ i, Set (α i)) :
+    (pi s fun i => if p i then t₁ i else t₂ i) =
+      pi ({ i ∈ s | p i }) t₁ ∩ pi ({ i ∈ s | ¬p i }) t₂ :=
+  by
+  ext f
+  refine' ⟨fun h => _, _⟩
+  ·
+    constructor <;>
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Prod.lean:765]:
```suggestion
  · constructor <;>
```

File: Mathlib/Data/Set/Prod.lean [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+
+theorem eval_image_univ_pi_subset : eval i '' pi univ t ⊆ t i :=
+  eval_image_pi_subset (mem_univ i)
+#align set.eval_image_univ_pi_subset Set.eval_image_univ_pi_subset
+
+theorem subset_eval_image_pi (ht : (s.pi t).Nonempty) (i : ι) : t i ⊆ eval i '' s.pi t := by
+  classical
+    obtain ⟨f, hf⟩ := ht
+    refine' fun y hy => ⟨update f i y, fun j hj => _, update_same _ _ _⟩
+    obtain rfl | hji := eq_or_ne j i <;> simp [*, hf _ hj]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Set/Prod.lean:822]:
```suggestion
  obtain ⟨f, hf⟩ := ht
  refine' fun y hy => ⟨update f i y, fun j hj => _, update_same _ _ _⟩
  obtain rfl | hji := eq_or_ne j i <;> simp [*, hf _ hj]
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**